subsys/net/lib/zzhc/zzhc.c

```diff
@@ -0,0 +1,639 @@
+		return -EILSEQ;
+	}
+	cptr++;
+
+	/* Swap bytes, including the extra check-digit */
+	for (i = 0; i <= ICCID_LEN; i += 2) {
+		c = cptr[i];
+		cptr[i] = cptr[i + 1];
+		cptr[i + 1] = c;
+	}
+	cptr[ICCID_LEN] = 0;
+
+	rc = memcmp(ctx->ueiccid, cptr, ICCID_LEN);
+	if (rc == 0) {
+		LOG_DBG("Identical ICCID.");
+	} else {
+		memcpy(ctx->ueiccid, cptr, ICCID_LEN);
+	}
+
+	return rc;
+}
+
+/**@brief Finite state machine */
+static int fsm(struct zzhc *ctx)
+{
+	int rc;
+
+	switch (ctx->state) {
+	case STATE_INIT:
+		LOG_DBG("STATE_INIT");
+
+		zzhc_sem_init(ctx->sem, 0, 0);
+		zzhc_register_handler(ctx, at_notif_handler);
+		ctx->state = STATE_WAIT_FOR_REG;
+		break;
+
+	case STATE_WAIT_FOR_REG:
+		LOG_DBG("STATE_WAIT_FOR_REG");
+
+		if (zzhc_at_cmd_xfer("AT+CEREG?", ctx->at_resp,
+			AT_RESPONSE_LEN) != 0) {
+			return -EIO;
+		}
+
+		rc = zzhc_get_at_param_short(ctx, ctx->at_resp, 1);
+		if (rc == NW_REGISTERED_HOME || rc == NW_REGISTERED_ROAM) {
+			ctx->registered = true;
+		} else {
+			zzhc_sem_wait(ctx->sem);
+		}
+
+		if (check_iccid(ctx) == 0 ||
+			check_simcard_whitelist(ctx) != 0) {
+			ctx->state = STATE_END;
+		} else {
+			ctx->state = STATE_START;
+		}
+		break;
+
+	case STATE_START:
+		LOG_DBG("STATE_START");
+
+		/* Compile HTTP request. */
+		rc = build_http_req(ctx);
+		if (rc <= 0) {
+			ctx->state = STATE_RETRY_TIMEOUT;
+			break;
+		}
+
+		/* Send. */
+		if (resolve_and_connect(ctx) != 0) {
+			ctx->state = STATE_RETRY_TIMEOUT;
+			break;
+		}
+
+		if (post_request_send(ctx) != 0) {
+			ctx->state = STATE_RETRY_TIMEOUT;
+			break;
+		}
+
+		ctx->rx_offset = 0;
+		ctx->delimiter = 0;
+		ctx->header_recvd = false;
+		ctx->state = STATE_HTTP_REQ_SENT;
+		break;
+
+	case STATE_HTTP_REQ_SENT:
+		LOG_DBG("STATE_HTTP_REQ_SENT");
+
+		rc = receive_response(ctx);
+		if (rc < 0) {
+			LOG_DBG("Error reading from socket, errno=%d", errno);
+			ctx->state = STATE_RETRY_TIMEOUT;
+			break;
+		}
+
+		if (rc == 0) {
+			LOG_DBG("Peer closed connection!");
+			ctx->state = STATE_HTTP_RESP_RECVD;
+			break;
+		}
+		break;
+
+	case STATE_HTTP_RESP_RECVD:
+		LOG_DBG("STATE_HTTP_RESP_RECVD");
+
+		/* Parse HTTP response. */
+		if (zzhc_check_http_payload(ctx)) {
+			/* Try to write ICCID to local storage. */
+			rc = zzhc_save_iccid(ctx->ueiccid, ICCID_LEN);
+			if (rc <= 0) {
+				LOG_DBG("zzhc_save_iccid() = %d", rc);
+			}
+
+			ctx->state = STATE_END;
+		} else {
+			ctx->state = STATE_RETRY_TIMEOUT;
+		}
+		break;
+
+	case STATE_RETRY_TIMEOUT:
+		LOG_DBG("STATE_RETRY_TIMEOUT");
+
+		/* Cleanup */
+		disconnect(ctx);
+		ctx->http_pkt_len = 0;
+		memset(ctx->http_pkt, 0, HTTP_PKT_LEN);
+
+		if (ctx->retry_cnt >= RETRY_CNT_MAX) {
+			ctx->state = STATE_END;
+			break;
+		}
+
+		zzhc_sleep(RETRY_TIMEOUT);
+		ctx->retry_cnt++;
+		ctx->state = (ctx->registered)
+			? STATE_START : STATE_WAIT_FOR_REG;
+		break;
+
+	case STATE_END:
+		LOG_DBG("STATE_END");
+
+		/* Clean-up */
+		disconnect(ctx);
+		zzhc_deregister_handler(ctx, at_notif_handler);
```

Comment:
Should you not deregister in any case when the state machine is finished in case an error occurs in a previous state?

Reply:
For any error happening in the state machine, the state `STATE_RETRY_TIMEOUT` will be reached. After `RETRY_CNT_MAX` (i.e. 3) times of any error, `STATE_END` will be reached. Such state refers to the end of execution.

The document from China Telecom states that "After execution, it should quit and not be persistent." This implies that `ctx` should be freed and `at_notif_handler` de-registered.